x/interchainstaking/ibc_module.go

```diff
@@ -209,6 +221,60 @@ func (im IBCModule) OnChanOpenAck(
 	return nil
 }
 
+func (im IBCModule) registerPerformanceAddress(
+	ctx sdk.Context,
+	portId,
+	connectionId,
+	chainId,
+	address string,
+	zone *types.RegisteredZone,
+) error {
+	zone.PerformanceAddress = &types.ICAAccount{Address: address, Balance: sdk.Coins{}, DelegatedBalance: sdk.NewCoin(zone.BaseDenom, sdk.ZeroInt()), PortName: portId}
+
+	// set withdrawal address if, and only if withdrawal address is already set
+	if zone.WithdrawalAddress != nil {
+		msg := distrTypes.MsgSetWithdrawAddress{DelegatorAddress: address, WithdrawAddress: zone.WithdrawalAddress.String()}
+		im.keeper.SubmitTx(ctx, []sdk.Msg{&msg}, zone.PerformanceAddress)
+	}
+
+	var cb keeper.Callback = func(k keeper.Keeper, ctx sdk.Context, response []byte, query icqtypes.Query) error {
+		zone, found := k.GetRegisteredZoneInfo(ctx, query.GetChainId())
+		if !found {
+			return fmt.Errorf("no registered zone for chain id: %s", query.GetChainId())
+		}
+
+		// initialize performance delegations
+		if err := k.InitPerformanceDelegations(ctx, zone, response); err != nil {
```

Comment:
This is being called every keyDelegations blocks - is this correct?

Reply:
No, it should only run until successful... (initial performance delegations)

1. Query balance
2. If sufficient, send performance delegations...
3. Then, stop.

From here the zone callbacks dealing with QueryDelegationTotalRewards call WithdrawDelegatorReward for the performance account of the zone ("same as if the delegator simply started a new delegation of the same value"); This action resets the performance scoring for the next epoch...

We have to deal with changes to the validator set, such as validators dropping off (or joining) the active set, validators being jailed, etc; I think it would be best to hook performance delegation management mechanisms into that logic for better overall clarity.